Step11-2 Michelle & Koon

leehwarang opened the pull request:
### 구현한 캐러셀 기능
- 무한 루프
- Pagination 

### 설계
- 클래스는 Carousel(Btn 포함)과 Pagination으로 분리하였습니다.
    - 이유는? Btn은 캐러셀에 필수적인 요소이지만, Pagination은 선택적 항목이고 수정 될 확률이 많다고 생각했습니다.

### 설계 과정 중 궁금한 점
- **Carousel의 인스턴스를 Pagination의 클래스에 주입하였습니다.** Pagination은 결국 Carousel의 컨텐츠를 변경하는 방법이기 때문에, Carousel의 프로퍼티와 메서드를 사용해야 했습니다.
-  다만, Pagination 에서 Carousel의 메서드를 사용할 때 기본 로직은 동일하지만 달라지는 부분이 있었습니다.
- 따라서 Carousel의 메서드에 `navoption = false` 을 dafault parameter로 주었고,  Pagination에서 이 메서드를 호출 시 navoption을 true로 넘기며 분기 처리를 하였습니다.

많은 고민을 하면서 코드를 작성 하였는데 이 설계 방식이 적절한지 알고 싶습니다.

### 참고 사항
- 현재 무한 루프 옵션은 true일 때만 동작합니다. 추후 애니메이션 적용시, 무한 루프 옵션이 false인 경우도 함께 구현하겠습니다.

leehwarang committed:
필요한 이미지 파일 저장

leehwarang committed:
feat 1차 마크업 진행 및 reset css 추가

leehwarang committed:
feat 1차 마크업에 스타일 적용

- reset.css와 style.css 파일 분리

leehwarang committed:
feat 2차 마크업 진행

- 버튼 추가

leehwarang committed:
feat 2차 마크업에 스타일 적용 및 케러셀 스크립트 추가

- prev, next 버튼 누을 때마다 좌,우로 이동 (무한 루핑 전)

leehwarang committed:
feat 캐러셀 무한루프 작동을 위한 로직 추가

- 각 캐러셀 아이템을 인덱스로 선택할 수 있도록 함
- 첫 번째와 마지막 캐러셀 아이템을 클론하여 처음과 끝에 추가

leehwarang committed:
feat 캐러셀 무한루프 기능 추가

leehwarang committed:
feat 네비게이션 클릭했을 때, 해당 인덱스의 캐러셀 아이템 보이게 하기

leehwarang committed:
refactor Carousel을 클래스 형태로 변경

leehwarang committed:
refactor Carousel 클래스 내부에 Btn에 관련된 로직도 추가

leehwarang committed:
refactor Nav를 클래스 형태로 변경

leehwarang committed:
style 파일명 수정

crongro commented on images/right-chevron.svg:
```diff
@@ -0,0 +1,37 @@
+<?xml version="1.0" encoding="iso-8859-1"?>
```
이건 뭐에욤

leehwarang replied:
svg 파일을 가져와서 사용한거라 내부 코드가 어떻게 되어있는지는 확인하지 않았었는데, utf-8 처럼 svg 파일을 인코딩 하는 방법 중 하나라고 하네요!

crongro commented on script.js:
```diff
@@ -0,0 +1,183 @@
+  }
+}
+
+class Carousel {
```
다음엔 한개의 파일에 하나의 Class만 존재하도록 하세요.
그렇게 프로그램 소스를 분리해야 수정할때 변경영향이 적어서 안전해요.

leehwarang replied:
넵! 각 클래스를 다른 파일로 분리하여 모듈로 생성하는 것 적용했습니다~!

crongro commented on script.js:
```diff
@@ -0,0 +1,183 @@
+
+  attachEventToPagination() {
+    this.navigation.addEventListener("click", e => {
+      let navIndex = e.target.getAttribute("nav-index");
```
data저장목적은 data attribute 라는 표준방법을 찾아서 적용해보세요.

kkw10 replied:
dataset을 사용해서 속성을 가져오는 방식으로 적용했습니다.
다만, dataset과 getAttribute의 차이점이 무엇인지는 궁금합니다.

crongro replied:
기능상으로 같고요,  
dataset 속성을 사용해서 쉽고 명시적으로, 접근할 수 있습니다.

crongro commented on script.js:
```diff
@@ -0,0 +1,183 @@
+class Pagination {
```
역할 분리는 좋은데요. 
carousel 모듈에 대한 dependency가 조금 높아보이는데요.
어려운 미션이겠지만 의존성을 조금이라도 낮출 수 있을지 더 고민해보면 좋겠어요.

crongro commented on script.js:
```diff
@@ -0,0 +1,183 @@
+
+class Carousel {
+  constructor(input, option) {
+    this.carousel = document.querySelector(input);
```
엘리먼트노드 이름이, 클래스이름과 같네요.  헷갈립니다 ㅎ

crongro commented on script.js:
```diff
@@ -0,0 +1,183 @@
+    this.carouselItems = this.carousel.children;
+    this.carouselWidth = this.carousel.offsetWidth;
+    this.carouselItemsInitCnt = this.carouselItems.length;
+    this.carouselItemsCnt = this.carouselItems.length;
```
모든 게 다 carousel 기반노드들이니까 이름에서 아예 carousel을 빼도 상관없을 거 같은데요.

crongro commented on script.js:
```diff
@@ -0,0 +1,183 @@
+  }
+
+  init() {
+    if (this.option.infinite) {
```
init 함수는 초기화 하는 그런 역할인거 같은데요.
하위함수 여러개를 만들어서, 컴포넌트가 초기화 될때 어떤 일을 하는 것인지 좀더 쉽게 알 수 있게 해주세요.

crongro commented on script.js:
```diff
@@ -0,0 +1,183 @@
+
+  appendCloneItem(parentNode, firstItem, lastItem) {
+    let firstCloneItem = firstItem.cloneNode(true);
+    let lastCloneItem = lastItem.cloneNode(true);
```
node를 clone해도 좋고, 
HTML에 미리 노드를 만들어 두어도 좋죠.

crongro commented on script.js:
```diff
@@ -0,0 +1,183 @@
+
+  setAttrToElement(elementName, attrName, navoption = false) {
+    let cnt = elementName.length;
+    for (let i = 0; i < cnt; i += 1) {
```
대부분의 for 문은 forEach나 for-of로~ 대체할 수 있을거에요.

leehwarang replied:
유사배열을 배열로 바꾸는 Array.from() 메서드를 배웠고, 변환 후 forEach로 수정하였습니다.

crongro commented on script.js:
```diff
@@ -0,0 +1,183 @@
+
+  getActiveItem() {
+    let active = document.querySelector(".active");
+    return active.getAttribute("data-index");
```
위에서 언급했듯이 data attribute를 잘 활용하기.

crongro commented on script.js:
```diff
@@ -0,0 +1,183 @@
+
+  updateActiveItem(activeIndex, moveValue, navoption = false) {
+    this.carouselItems[activeIndex].classList.remove("active");
+    if (navoption === true) {
```
=== true 는  불필요~

crongro commented on script.js:
```diff
@@ -0,0 +1,183 @@
+  }
+
+  updateCloneActiveItem(newActiveItem, value) {
+    newActiveItem.classList.remove("active");
```
active 라는 클래스명이 바뀌면, 코드의 이곳저곳을 다 수정해야 할 거 같아요.
어떻게 해야 변경에 영향을 덜 받게 할 수 있을까요?

crongro commented on script.js:
```diff
@@ -0,0 +1,183 @@
+      let newActiveItem = this.updateActiveItem(activeIndex, 1);
+      this.moveCarousel(-1);
+
+      if (this.option.infinite) {
```
true조건보다 false조건을 체크해서 return 해버리면 중첩된 if문이 하나 줄거 같아요.

crongro commented on script.js:
```diff
@@ -0,0 +1,183 @@
+    this.carouselWrap = this.carousel.parentNode;
+    this.carouselItems = this.carousel.children;
+    this.carouselWidth = this.carousel.offsetWidth;
+    this.carouselItemsInitCnt = this.carouselItems.length;
```
Cnt 라는 이름이 count 죠? 그냥 count라고 다 써도 될거 같아요.
그것보다 Cnt 앞부분의 이름을 더 줄이는 게 좋을 거 같은데요.

crongro reviewed (changes requested):
고민을 좀 드리고 싶네요. 의존성을 줄이는 것도 생각해보시고요.
객체를 주입하지 않고 함수를 넘겨주는 것도 괜찮을거 같아보이기도 하고요.
그외 다른 방법은 너무 어려울 듯 (제 3자를 만들어서 중재하는 클래스를 둔다던가 ㅎ)

나머지 수정해서 다시 push해주세요~

leehwarang committed:
refactor 각 클래스를 다른 파일로 분리하여 모듈로 생성

leehwarang committed:
refactor getAttribute()를 dataset으로 가져오는 표준 방법으로 변경

leehwarang committed:
refactor

- Carousel의 생성자에 선언한 변수명 간단히 수정
- 유사 배열을 순회하던 for문을 배열로 변환하여 forEach로 수정
- Carousel의 생성자에서 사용하던 변수를 지역변수로 이동
- 매직 스트링 제거

leehwarang committed:
refactor init() 함수의 로직 분리

leehwarang committed:
feat 애니메이션 추가

- 캐러셀 컨텐츠 이동시 움직이는 과정이 보이는 애니메이션 추가
- 캐러셀 컨텐츠 이동시 마다 해당하는 nav가 커지는 애니메이션 추가

leehwarang committed:
refactor 불필요한 코드 삭제

leehwarang commented:
### **https://kkw10.github.io/**
주신 피드백 적용 및 애니메이션 구현 추가했습니다. 
다만..  Carousel과 Pagination의 의존성이 높아서 지금처럼 서로 인스턴스를 주고 받는게 아닌 옵저버 패턴등을 적용해보라고 하셨는데, 아직 이 부분을 이해하지 못해서 적용하지 못했습니다.
개인적으로 학습 후 구현해보도록 하겠습니다.

crongro commented on pagination.js:
```diff
@@ -0,0 +1,37 @@
+  }
+
+  init() {
+    let navoption = true;
```
변경이 되는 변수가 아니면 let보다는 const겠죠.

crongro commented on app.js:
```diff
@@ -0,0 +1,12 @@
+  infinite: true
+});
+
+const pagination = new Pagination(carousel);
```
pagination을 초기화할때에 carousel을 넣는데,
carousel은 초기화할때 넣지 않고 init을 할때 넣네요.

crongro commented on app.js:
```diff
@@ -0,0 +1,12 @@
+import Carousel from "./carousel.js";
+import Pagination from "./pagination.js";
+
+const carousel = new Carousel(".carousel", {
```
carousel 과 pagination의 의존관계를 느슨하게 하는 건 쉬운건 아니에요. observer패턴이 해결책이 되지만 아직  적용하지 않아도 됩니다. 
또 다른 방법은 carousel과 pagination은 view역할만해서 렌더링하는 일만 하도록 하고, 
두 개가 서로 호출하며 의존하는 부분의 표현은 controller라고 하는 중간계층의 클래스를 만들어서 그곳에서 표현할 수 있을거에요.
아마 서로 호출하는 부분이 이벤트리스너 안에서 이뤄지는 일들이겠죠.

crongro commented on carousel.js:
```diff
@@ -0,0 +1,179 @@
+  setCarouselItem() {
+    this.appendCloneItem(
+      this.carousel,
+      this.items[0],
```
개인적으로 그다지 좋아하진 않는데요. 배열의 0, 1 이런식으로 표현하면 이게 무엇인지 알수가 없죠. 리스트 내용이 추가되면 깨질 수 있는 코드기도 하고요.

crongro commented on carousel.js:
```diff
@@ -0,0 +1,179 @@
+    this.addElementToClass(this.carousel, "flex");
+  }
+
+  setAttrToElement(elementName, attrName, navoption = false) {
```
navoption도 카멜표기법하면 좋았을 듯

crongro commented on carousel.js:
```diff
@@ -0,0 +1,179 @@
+    } else {
+      this.currentPointer += moveValue * this.width;
+    }
+    this.carousel.style.transition = `all 1s`;
```
이부분이 반복적인 거 같네요.
함수로 만들기를 고민할 수 있겠죠.

```
 this.carousel.style.transition = `all 1s`;
this.carousel.style.transform = `translateX(${this.currentPointer}px)`;
```

crongro commented on carousel.js:
```diff
@@ -0,0 +1,179 @@
+  }
+
+  attchEventToBtn() {
+    this.prevBtn.addEventListener("click", () => {
```
event listener 는 별도로 분리해서 선언하는 게  더 좋죠. 
event listener 를 직접 호출하면서 테스트코드를 구현하기도 좋고요. 그럴가망성은 낮아도 다른데서 콜백함수를 재사용할 수도 있을테고요.

crongro commented on carousel.js:
```diff
@@ -0,0 +1,179 @@
+
+      if (
+        this.option.infinite &&
+        newActiveItem.classList.contains(this.config.clone)
```
contains 메서드의 browser compatibility 를 한번 조사해볼래요? 
이런거 자주 들여다 보는 습관이 필요하실거에요.

crongro commented on carousel.js:
```diff
@@ -0,0 +1,179 @@
+      ) {
+        this.updateCloneActiveItem(newActiveItem, 1);
+        let newPointerValue = this.width;
+        setTimeout(() => this.moveCloneCarousel(newPointerValue), 1000);
```
1000은 매직넘버라...함수내에서 지우면 좋겠어요.

crongro commented on pagination.js:
```diff
@@ -0,0 +1,37 @@
+  }
+
+  attachEventToPagination() {
+    this.navList.addEventListener("click", e => {
```
e 대신에 {target} 으로 표현하는 것도 좋죠.
(destructuring)

crongro commented on pagination.js:
```diff
@@ -0,0 +1,37 @@
+      let currentactiveIndex = this.carousel.getActiveItem();
+
+      this.scaleUp(navIndex - 1);
+      this.carousel.updateActiveItem(currentactiveIndex, navIndex, true);
```
마지막 true 인자는 어떤 것을 의미하는 것인지 알수가 없죠. 뭐가 true일까? 생각이 들어요.
그럴때 객체를 활용해서 인자를 전달하는 것이 좋은 방법이죠.

crongro reviewed (approved):
컨트롤러나 옵저버패턴을 적용할 필요는 아직 없습니다.
더 복잡한 UI구조를 개발할때 적용할 기회가 있을거에요.

두 개의 클래스로 잘 나눴고요. 리뷰몇개 남기니 참고하세요.
수고했네요.

crongro commented:
Carousel의 인스턴스를 Pagination의 클래스에 주입하였습니다. Pagination은 결국 Carousel의 컨텐츠를 변경하는 방법이기 때문에, Carousel의 프로퍼티와 메서드를 사용해야 했습니다.
다만, Pagination 에서 Carousel의 메서드를 사용할 때 기본 로직은 동일하지만 달라지는 부분이 있었습니다.
따라서 Carousel의 메서드에 navoption = false 을 dafault parameter로 주었고, Pagination에서 이 메서드를 호출 시 navoption을 true로 넘기며 분기 처리를 하였습니다.

=> 네 그런 flag를 두고 분기처리하도록 하는  것도 가능하겠죠. 이상하진 않습니다. 다른방법은 비슷한 메서드를 하나 더 만드는 것이겠죠.
